File: js/cracking_coding_interview/linked_lists/removeDups.js
//doubly linked
function removeDup (head) {
  const seen = new Set();
  let currNode = head;
  while (currNode !== null) {
    if (seen.has(currNode.val)) {
      //remove node
      currNode.prev.next = currNode.next;
      currNode.next.prev = currNode.prev;
    } else {
      seen.add(currNode.val);
    }
    currNode = currNode.next;
  }
}

//singly linked
function removeDup2 (head) {
  const seen = new Set();
  let currNode = head;
  let prev = new Node(null);
  while (currNode !== null) {
    if (seen.has(currNode.val)) {
      prev.next = currNode.next;
    } else {
      seed.add(currNode.val);
      prev = currNode;
    }
    currNode = currNode.next;
  }
}

//no buffer allowed
function removeDup3 (head) {
  const currNode = head;
  while (currNode !== null) {
    let runner = current;
    while (runner.next !== null) {
      if (runner.next.val === currNode.val) {
        runner.next = runner.next.next;
      } else {
        runner = runner.next;
      }
    }
    currNode = currNode.next;
  }
}

File: ruby/cracking_coding_interview/linked_lists/remove_dups.rb
# Write code to remvoe dupicates from an unsorted linked list.
# Follow up: How would you solve this problem if a temporary buffer is not allowed?

# time: n
# space: n
require 'set'
def remove_dups1(head)
  node = head
  return node if node.val.nil?

  seen = Set.new(node.val)

  while !node.next.val.nil?
    if seen.include?(node.next.val)
      node.next = node.next.next
    else
      seen << node.next.val
    end
    node = node.next
  end

  head
end

# time: n^2
# space: 1
def remove_dups2(head)
  node1 = head

  while !node1.val.nil?
    prev = node1
    node2 = node1.next
    while !node2.val.nil?
      if node1.val == node2.val
        prev.next = node2.next
      else
        prev = node2
      end
      node2 = node2.next
    end
    node1 = node1.next
  end

  head
end
